### PR TITLE
Test run deletion feature added

### PR DESCRIPTION
**Pull request by Ark2307**

### CODE REVIEW

File: apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java [hunk 1]
```diff
@@ -1968,5 +1972,23 @@ public void run() {
             }
         }, 0, 1, UsageUtils.USAGE_CRON_PERIOD);
     }
+
+    public void deleteTestRunsScheduler(){
+        scheduler.scheduleAtFixedRate(new Runnable() {
+            public void run(){
+                List<DeleteRunsInfo> deleteRunsInfos = DeleteRunsInfoDao.instance.findAll(Filters.empty());
+                if(deleteRunsInfos.size() > 0){
```
Comment (ayushaga14) [inline apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java:1980]: > 0 check is redundant

File: apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java [hunk 2]
```diff
@@ -1968,5 +1972,23 @@ public void run() {
             }
         }, 0, 1, UsageUtils.USAGE_CRON_PERIOD);
     }
+
+    public void deleteTestRunsScheduler(){
```
Comment (ayushaga14) [inline apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java:1976]: do not keep the cron in this file

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+    public static final DeleteRunsInfoDao instance = new DeleteRunsInfoDao();
+
+    private static Map<String, Boolean> deleteTestRunsCollections(List<ObjectId> testRunIds,
+            List<ObjectId> latestSummaryIds, List<Integer> testConfigIds,
+            Map<String, Boolean> testingCollectionsDeleted) {
+                
+        if (!testingCollectionsDeleted.getOrDefault("testing_run_result", false)) {
+            TestingRunResultDao.instance.deleteAll(Filters.in(TestingRunResult.TEST_RUN_RESULT_SUMMARY_ID, latestSummaryIds));
+            testingCollectionsDeleted.put("testing_run_result", true);
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:27]: for each query, verify if index is created or not on the specific field

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+import com.akto.dao.AccountsContextDao;
+import com.akto.dao.testing_run_findings.TestingRunIssuesDao;
+import com.akto.dto.test_run_findings.TestingRunIssues;
+import com.akto.dto.testing.DeleteRunsInfo;
+import com.akto.dto.testing.TestingRunResult;
+import com.mongodb.client.model.Filters;
+import com.mongodb.client.model.Updates;
+
+public class DeleteRunsInfoDao extends AccountsContextDao<DeleteRunsInfo> {
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:17]: DeleteRunsInfoDao is a bad name, rename this to something understandable

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    public static final DeleteRunsInfoDao instance = new DeleteRunsInfoDao();
+
+    private static Map<String, Boolean> deleteTestRunsCollections(List<ObjectId> testRunIds,
+            List<ObjectId> latestSummaryIds, List<Integer> testConfigIds,
+            Map<String, Boolean> testingCollectionsDeleted) {
+                
+        if (!testingCollectionsDeleted.getOrDefault("testing_run_result", false)) {
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:25]: are all these 5 queries dependent on each other? if not then execute them parallely

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                    .deleteAll(Filters.in("_id", latestSummaryIds));
+            testingCollectionsDeleted.put("testing_run_result_summaries", true);
+        }
+
+        if (!testingCollectionsDeleted.getOrDefault("testing_run_config", false)) {
+            TestingRunConfigDao.instance.deleteAll(Filters.in("_id", testConfigIds));
+            testingCollectionsDeleted.put("testing_run_config", true);
+        }
+
+        if (!testingCollectionsDeleted.getOrDefault("testing_run", false)) {
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:47]: testRunIds is being passed as [] if summary is deleted. Can we avoid executing this query in that case?

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            TestingRunDao.instance.deleteAll(Filters.in("_id", testRunIds));
+            testingCollectionsDeleted.put("testing_run", true);
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:49]: did you handle what we'll show if all summary ids have been deleted?

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        List<ObjectId> latestSummaryIds = deleteRunsInfo.getLatestTestingSummaryIds();
+        List<Integer> testConfigIds = deleteRunsInfo.getTestConfigIds();
+
+        testingCollectionsDeleted = deleteTestRunsCollections(testRunIds, latestSummaryIds, testConfigIds,
+                testingCollectionsDeleted);
+
+        Bson filter = Filters.in(DeleteRunsInfo.TEST_RUNS_IDS, testRunIds);
+        deleteRunsInfo.setTestingCollectionsDeleted(testingCollectionsDeleted);
+        if (DeleteRunsInfoDao.instance.isTestRunDeleted(deleteRunsInfo)) {
+            DeleteRunsInfoDao.instance.getMCollection().deleteOne(filter);
```
Comment (ayushaga14) [inline libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:67]: confirm once with ankush whetheer we should delete this document as well

**Review by ayushaga14 — CHANGES_REQUESTED** [review]
fix comments

File: apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java (hunk 1 above)
Comment (Ark2307) [reply apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java:1980]: Removed it, but still checking for null .

File: apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java (hunk 2 above)
Comment (Ark2307) [reply apps/dashboard/src/main/java/com/akto/listener/InitializerListener.java:1976]: Created a folder crons in utils

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java (hunk 5 above)
Comment (Ark2307) [reply libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:25]: There's a tree in dependency. Our testing has a tree like structure hence not fired parallely

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java (hunk 7 above)
Comment (Ark2307) [reply libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:49]: yess in the ui, an empty screen will come up

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java (hunk 4 above)
Comment (Ark2307) [reply libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:17]: is deleteTestRuns a good name?

File: libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java (hunk 6 above)
Comment (Ark2307) [reply libs/dao/src/main/java/com/akto/dao/testing/DeleteRunsInfoDao.java:47]: For the cases of Ci/Cd and recurring, we dont want to delete test runs and its config. So in that case we use summaryIds to delete. But to use the same function, i passed testrunids and configids as empty. so there'll be no document found in mongo

**Review by ayushaga14 — APPROVED** [review]

**Review by ayushaga14 — APPROVED** [review]

**Pull request merged**